cmd/solana_exporter/slots.go

```diff
@@ -91,96 +117,82 @@ func (c *solanaCollector) WatchSlots() {
 		epochFirstSlot.Set(float64(firstSlot))
 		epochLastSlot.Set(float64(lastSlot))
 
-		// Check whether we need to fetch a new leader schedule
-		if epochNumber != info.Epoch {
-			klog.Infof("new epoch at slot %d: %d (previous: %d)", firstSlot, info.Epoch, epochNumber)
-
-			epochSlots, err = c.fetchLeaderSlots(firstSlot)
-			if err != nil {
-				klog.Errorf("failed to request leader schedule, retrying: %v", err)
-				continue
-			}
-
-			klog.V(1).Infof("%d leader slots in epoch %d", len(epochSlots), info.Epoch)
-
-			epochNumber = info.Epoch
-			klog.V(1).Infof("we're still in epoch %d, not fetching leader schedule", info.Epoch)
-
-			// Reset watermark to current offset on new epoch (we do not backfill slots we missed at startup)
-			watermark = info.SlotIndex
-		} else if watermark == info.SlotIndex {
+		if watermark == info.AbsoluteSlot {
 			klog.Infof("slot has not advanced at %d, skipping", info.AbsoluteSlot)
 			continue
 		}
 
 		klog.Infof("confirmed slot %d (offset %d, +%d), epoch %d (from slot %d to %d, %d remaining)",
-			info.AbsoluteSlot, info.SlotIndex, info.SlotIndex-watermark, info.Epoch, firstSlot, lastSlot, lastSlot-info.AbsoluteSlot)
+			info.AbsoluteSlot, info.SlotIndex, info.AbsoluteSlot-watermark, info.Epoch, firstSlot, lastSlot, lastSlot-info.AbsoluteSlot)
 
-		// Get list of confirmed blocks since the last request. This is totally undocumented, but the result won't
-		// contain missed blocks, allowing us to figure out block production success rate.
-		rangeStart := firstSlot + watermark
-		rangeEnd := firstSlot + info.SlotIndex - 1
-
-		ctx, cancel = context.WithTimeout(context.Background(), httpTimeout)
-		cfm, err := c.rpcClient.GetConfirmedBlocks(ctx, rangeStart, rangeEnd)
+		last, err := updateCounters(c, currentEpoch, watermark, nil)
 		if err != nil {
-			klog.Errorf("failed to request confirmed blocks at %d, retrying: %v", watermark, err)
-			cancel()
+			klog.Info(err)
 			continue
 		}
-		cancel()
-
-		klog.V(1).Infof("confirmed blocks: %d -> %d: %v", rangeStart, rangeEnd, cfm)
+		watermark = last
+	}
+}
 
-		// Figure out leaders for each block in range
-		for i := watermark; i < info.SlotIndex; i++ {
-			leader, ok := epochSlots[i]
-			abs := firstSlot + i
-			if !ok {
-				// This cannot happen with a well-behaved node and is a programming error in either Solana or the exporter.
-				klog.Fatalf("slot %d (offset %d) missing from epoch %d leader schedule",
-					abs, i, info.Epoch)
-			}
+func updateCounters(c *solanaCollector, epoch, firstSlot int64, lastSlotOpt *int64) (int64, error) {
```

Comment:
Can the first argument here not be an `RPCClient` instead? As only `c.rpcClient` is used in the function

Reply:
done